### PR TITLE
Avereha/notifications

### PR DESCRIPTION
**Pull request by avereha**
Notify on unconfirmed commands.
Does not beep on notifications, because we are missing options to disable that beep.

### CODE REVIEW

File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/comm/OmnipodDashBleManagerImpl.kt [hunk 1]
```diff
@@ -125,7 +125,7 @@ class OmnipodDashBleManagerImpl @Inject constructor(
             for (i in 1..MAX_NUMBER_OF_CONNECTION_ATTEMPTS) {
                 try {
                     // wait i * CONNECTION_TIMEOUT
-                    conn.connect(i)
+                    conn.connect(4)
```
Comment (bartsopers) [inline omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/comm/OmnipodDashBleManagerImpl.kt:128]: Nit: I think it would be good to make this a constant

File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/pod/state/OmnipodDashPodStateManager.kt [hunk 2]
```diff
@@ -89,6 +96,13 @@ interface OmnipodDashPodStateManager {
 
     fun createLastBolus(requestedUnits: Double, historyId: String, bolusType: DetailedBolusInfo.BolusType)
     fun markLastBolusComplete(): LastBolus?
+    fun onStart()
+    /*
+    This is called only:. It overwrites activationStatus
+       - when activation was interrupted(application crash, killed, etc)
+       - after getPodStatus was successful(we have an up-to-date podStatus)
+     */
+    fun recoverActivationFromPodStatus(): String?
```
Comment (bartsopers) [inline omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/pod/state/OmnipodDashPodStateManager.kt:105]: The implementation always seems to return null. Shouldn't we use a `void` return type here?

File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/pod/state/OmnipodDashPodStateManagerImpl.kt [hunk 3]
```diff
@@ -379,6 +466,9 @@ class OmnipodDashPodStateManagerImpl @Inject constructor(
         podState.lastUpdatedSystem = System.currentTimeMillis()
         podState.lastStatusResponseReceived = SystemClock.elapsedRealtime()
         updateLastBolusFromResponse(response.bolusPulsesRemaining)
+        if (podState.activationTime == null) {
+            podState.activationTime = System.currentTimeMillis() - (response.minutesSinceActivation * 60000)
```
Comment (bartsopers) [inline omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/pod/state/OmnipodDashPodStateManagerImpl.kt:470]: Nit: you could use 60_000 for readability :-)

File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/ui/OmnipodDashOverviewFragment.kt [hunk 4]
```diff
@@ -291,52 +268,36 @@ class OmnipodDashOverviewFragment : DaggerFragment() {
...
+                        Color.WHITE
+                    }
+                } ?: Color.WHITE
+            )
 
             // Update Pod expiry time
-            val expiresAt = getExpiryAt()
-            if (expiresAt is Nothing) {
+            val expiresAt = podStateManager.expiry
+            if (expiresAt == null) {
```
Comment (bartsopers) [inline omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/ui/OmnipodDashOverviewFragment.kt:287]: nit: can we use `expiresAt.let {
// ...
} ?: ...`
here?

File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/ui/OmnipodDashOverviewFragment.kt [hunk 5]
```diff
@@ -564,7 +523,6 @@ class OmnipodDashOverviewFragment : DaggerFragment() {
     private fun updateRefreshStatusButton() {
         buttonBinding.buttonRefreshStatus.isEnabled =
             podStateManager.isUniqueIdSet &&
-            podStateManager.activationProgress.isAtLeast(ActivationProgress.PHASE_1_COMPLETED) &&
```
Comment (bartsopers) [inline omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/ui/OmnipodDashOverviewFragment.kt:567]: Why did you remove this condition? AFAIK, we do a GetStatus request when the user presses the refresh button, and the Pod doesn't respond to these before ActivationProgress.PHASE_1_COMPLETED

File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/pod/state/OmnipodDashPodStateManagerImpl.kt [hunk 6]
```diff
@@ -513,6 +597,7 @@ class OmnipodDashPodStateManagerImpl @Inject constructor(
         var ltk: ByteArray? = null
         var eapAkaSequenceNumber: Long = 1
         var bolusPulsesRemaining: Short = 0
+        var timeZone: String = "" // TimeZone ID (e.g. "Europe/Amsterdam")
```
Comment (bartsopers) [inline omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/pod/state/OmnipodDashPodStateManagerImpl.kt:600]: Can't we store this as a TimeZone?

File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/ui/DashPodManagementActivity.kt [hunk 7]
```diff
@@ -125,8 +125,8 @@ class DashPodManagementActivity : NoSplashAppCompatActivity() {
 
         binding.buttonActivatePod.isEnabled = podStateManager.activationProgress.isBefore(ActivationProgress.COMPLETED)
         binding.buttonDeactivatePod.isEnabled =
-            podStateManager.activationProgress.isAtLeast(ActivationProgress.PHASE_1_COMPLETED) ||
-                podStateManager.podStatus == PodStatus.ALARM
+            podStateManager.ltk != null ||
```
Comment (bartsopers) [inline omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/ui/DashPodManagementActivity.kt:128]: AFAIK, it's not possible to send a DeactivatePodCommand before ActivationProgress.PHASE_1_COMPLETED

**Review by bartsopers — APPROVED** [review]

File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/pod/state/OmnipodDashPodStateManagerImpl.kt (hunk 6 above)
Comment (vanelsberg) [reply omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/pod/state/OmnipodDashPodStateManagerImpl.kt:600]: Sure, but (as this is a joda.time object?) it does not serialize w/o custom serialisation.
Fixed as TZ name string for now as without pod state could not be persisted. 
(TZ implementation can be done at a later time?)

File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/ui/OmnipodDashOverviewFragment.kt (hunk 4 above)
Comment (vanelsberg) [reply omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/ui/OmnipodDashOverviewFragment.kt:287]: Better (coder had little experience with Kotlin to find out how to use the construct)

File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/comm/OmnipodDashBleManagerImpl.kt (hunk 1 above)
Comment (avereha) [reply omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/comm/OmnipodDashBleManagerImpl.kt]: Done.
This whole section might go away in https://github.com/0pen-dash/AndroidAPS/pull/56 where I'm trying to implement `connect`/`disconnect` from the Pump interface. 
Not sure if this idea will work or not, still playing with it.

File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/pod/state/OmnipodDashPodStateManagerImpl.kt (hunk 3 above)
Comment (avereha) [reply omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/pod/state/OmnipodDashPodStateManagerImpl.kt:471]: Done

File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/pod/state/OmnipodDashPodStateManager.kt (hunk 2 above)
Comment (avereha) [reply omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/pod/state/OmnipodDashPodStateManager.kt:106]: It returns "busy" on `ENGAGING_CLUTCH_DRIVE` and `PRIMING`.
We are not using the return value yet. The idea is to "wait" and not continue the activation process in one of those two states.

File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/ui/OmnipodDashOverviewFragment.kt (hunk 4 above)
Comment (avereha) [reply omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/ui/OmnipodDashOverviewFragment.kt:287]: Will do in a future PR where we also fix https://github.com/0pen-dash/AndroidAPS/issues/59

File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/pod/state/OmnipodDashPodStateManagerImpl.kt (hunk 6 above)
Comment (avereha) [reply omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/pod/state/OmnipodDashPodStateManagerImpl.kt:601]: We tried this and deserialization if pod state failed.

File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/ui/OmnipodDashOverviewFragment.kt (hunk 5 above)
Comment (avereha) [reply omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/ui/OmnipodDashOverviewFragment.kt:567]: I tested and it seems to work ok. 

This is the bug that started all these changes: start activation, kill the app in the middle. One of the commands succeeded(I think it was the `setUniqueId` command), but the pod state was not updated yet. When I tried to continue the activation, got a screamer because that pod state was out of sync with the activation progress that we saved.

The fix I was thinking to implement was to call `getPodStatus` on startup and update the activation progress based on this status. For now, I just make the button visible and refresh the status by hand before I continue activation.

File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/ui/DashPodManagementActivity.kt (hunk 7 above)
Comment (avereha) [reply omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/ui/DashPodManagementActivity.kt:128]: I did not test this yet because each test will lose one pod. :(

The main issue I was trying to solve was that I wanted to have a way to reset pod state if activation fails in the middle. 

When I noticed the bug described here https://github.com/0pen-dash/AndroidAPS/pull/50#discussion_r667457543 I got to a state where I was not able to do anything(continue activation or discard pod). 

The worst thing that could happen if deactivation will not work is that we will get an error and then we will at least be able to discard the pod.

Opened https://github.com/0pen-dash/AndroidAPS/issues/60 so we don't forget about this
